src/python/pants/backend/tools/workunit_logger/rules.py

```diff
@@ -43,6 +43,16 @@ def just_dump_map(workunits_map):
     ]
 
 
+def pass_through_unserializable_metadata(obj):
+    """Most of the metadata in workunit objects is json serializable, but it is not guaranteed that
+    all of it is.
+
+    This function is intended to be used as a `default` encoder to json.dumps to drop a minimal stub
+    encoding with the name instead of throwing a TypeError and halting the entire workunit logging.
+    """
+    return {"name": obj.__class__.__name__, "json_serializable": False}
```

Comment:
(Neither of these are blocking a merge, e.g. merging this ASAP, but continuing the discussion and (if required) iterating in a follow-up PR is fine by me.)

What do you think about potentially also including `str(obj)` or `repr(obj)`? For many objects this is likely to be useful for humans.

```suggestion
    return {"name": obj.__class__.__name__, "json_serializable": False, "str": str(obj)}
```

And/or, should we flag this for users to fix with a warning? E.g.

```python
@memoized
def _warn_about_unserializable_type(ty: type):
    logger.warning("Value of type %s appears in workunit metadata, but is not JSON serializeable, and thus will not be included in workunit logging output. If this is in a custom plugin, consider fixing it; if this is in a builtin backend, please file an issue <https://github.com/pantsbuild/pants/issues/new/choose>", ty.__qualname__)

...

def pass_through_unserializable_metadata(obj):
    ...
    _warn_about_unserializable_type(type(obj))
    ...
```

(`@memoized` to ensure the warning is only shown once per type, to avoid massive output spam.)